### PR TITLE
1751 trace multiple transactions in a single block

### LINKED ISSUE #1750: Skaled return an error if block contains more than 1 transaction from different users for debug traceBlockByNumber
**Preconditions**
Active chains
Archival node
Deployed contract with erc20 tokens

**Versions**
skalenetwork/schain:3.17.2-develop.2-historic

**Steps to reproduce**

1. Send several transactions into the same block from different users 
2. Send debug.traceBlockByNumber method 

**Expected state:**
Skaled should return the tracing result by executing all transactions in the block specified by number with a tracer 

**Actual state:**
Skaled returns the error:
```
> debug.traceBlockByNumber('122177', {"tracer": "callTracer"})
Error: vector::_M_range_check: __n (which is 0) >= this->size() (which is 0)
    at web3.js:3143:20
    at web3.js:6347:15
    at web3.js:5081:36
    at <anonymous>:1:1

```

### LINKED ISSUE #1751: Skaled return an error if block contains more than 1 transaction from single user for debug traceBlockByNumber
**Preconditions**
Active chains
Archival node
Deployed contract with erc20 tokens

**Versions**
skalenetwork/schain:3.17.2-develop.2-historic

**Steps to reproduce**

1. Send several transactions into the same block from 1 user
2. Send debug.traceBlockByNumber method

**Expected state**:
Skaled should return the tracing result by executing all transactions in the block specified by number with a tracer

**Actual state**:
Skaled returns the error:
```
> debug.traceBlockByNumber('0x1dcc5', {"tracer": "callTracer"})
Error: State check failed::m_topFunctionCall /home/gha/actions-runner-1/_work/skaled/skaled/libhistoric/AlethStandardTrace.cpp:354
    at web3.js:3143:20
    at web3.js:6347:15
    at web3.js:5081:36
    at <anonymous>:1:1

{
  author: "0xd2ba3e0000000000000000000000000000000000",
  boundary: "0x0000000000000000000000000000000000000000000000000000000000000000",
  difficulty: 0,
  extraData: "0x736b616c65",
  gasLimit: 268435455,
  gasUsed: 106000,
  hash: "0x5d58670136d034b8598d10dcf77b9c11a81db21143946f4956407a5c2c2af7cb",
  logsBloom: "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  miner: "0xd2ba3e0000000000000000000000000000000000",
  mixHash: "0x0000000000000000000000000000000000000000000000000000000000000000",
  nonce: "0x0000000000000000",
  number: 122053,
  parentHash: "0xb50016280550d3a235e2a69701b18eb9d59b5c5901b7cde97a3bbfb61ac9e354",
  receiptsRoot: "0xbf3582ecd042451d77a3e737882a790c96b287b9a96965330d7af0fc32e652ff",
  seedHash: "0xb903bd7696740696b2b18bd1096a2873bb8ad0c2e7f25b00a0431014edb3f539",
  sha3Uncles: "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
  size: 675,
  stateRoot: "0x10082ac6d6578a7bb98227684ecd99753eea170eda7ada82144def92709620ff",
  timestamp: 1702405860,
  totalDifficulty: 0,
  transactions: ["0x0801c1ca24dee92ddef3594205f1cda1895e21312d285c9557b179c1d8c81fb3", "0x7d7a60dee54595e288ea1880ec107a811a3f24444ff804950af45ca04f32f90f"],
  transactionsRoot: "0x4e3b6310f238729194ab017e99b0e23aeb722f8c172f6af078ef691dead96086",
  uncles: []
}


```

```[tasklist]
### Tasks
- [ ] https://github.com/skalenetwork/skaled/issues/1750
```

### LINKED ISSUE #1753: prestateTracer, callTracer, 4byteTracer, noopTracer return the same answer for debug_traceTransaction JSON-RPC call
**Used calls** 
curl http://10.3.155.147:10131 -X POST -H "Content-Type: application/json" --data '{"method":"debug_traceTransaction","params":["0x1d2b8062fd975bc3f718beb9e391ea35a10e950fb9b94857a524364b70f2b317",{"tracer":"callTracer","tracerConfig":{"withLog":true}}],"id":1,"jsonrpc":"2.0"}' 
curl http://10.3.155.147:10131 -X POST -H "Content-Type: application/json" --data '{"method":"debug_traceTransaction","params":["0x1d2b8062fd975bc3f718beb9e391ea35a10e950fb9b94857a524364b70f2b317",{"tracer":"4byteTracer"}],"id":1,"jsonrpc":"2.0"}' 
curl http://10.3.155.147:10131 -X POST -H "Content-Type: application/json" --data '{"method":"debug_traceTransaction","params":["0x1d2b8062fd975bc3f718beb9e391ea35a10e950fb9b94857a524364b70f2b317",{"tracer":"prestateTracer"}],"id":1,"jsonrpc":"2.0"}'
curl http://10.3.155.147:10131 -X POST -H "Content-Type: application/json" --data '{"method":"debug_traceTransaction","params":["0x1d2b8062fd975bc3f718beb9e391ea35a10e950fb9b94857a524364b70f2b317",{"tracer":"noopTracer"}],"id":1,"jsonrpc":"2.0"}' 

**Preconditions**
Active chains
Archival node
Address with Sfuel and with Whitelist deployer role
Contract to deploy: [Example ERC20](https://github.com/skalenetwork/test-tokens/blob/develop/contracts/ERC20Custom.sol)

**Versions**
skalenetwork/schain:3.17.2-develop.2-historic

**Steps to reproduce**
1. Send transaction contract deployment to schain via any tool ( Hardhat, web3py, Remix)
2. Wait until the transaction is mined with status 0x1 (success) and save the transaction hash
3. Send 4 debug_traceTransaction methods with tx_hash from step 2  with every Geth Tracers 
```
    gethTracers = ["callTracer", "4byteTracer", "prestateTracer", "noopTracer"]
```

**Expected state:**
The prestateTracer returns the accounts necessary to execute a given transaction
callTracer -  tracks all the call frames executed during a transaction, including depth 0.
4byteTracer -  collects the function selectors of every function executed in the lifetime of a transaction, along with the size of the supplied call data. 
noopTracer - should return an empty object and is only meant for testing the setup

**Actual state:**
Skaled returns the same output for all 4 traces of a given transaction that is valid only for the callTracer argument

**Log output 4 times for each call**
```
{"id":1,"jsonrpc":"2.0","result":{"from":"0xb04a7f2cd74eb624df4bd8bed8f7034b5aed89c0","gas":"0x05ab9e","gasUsed":"0x06fc5a","output":"0x608060405234801561001057600080fd5b50600436106100625760003560e01c806306fdde031461006757806318160ddd1461008557806340c10f191461009c57806356b8c724146100b157806370a08231146100d457806395d89b41146100f4575b600080fd5b61006f6100fc565b60405161007c9190610483565b60405180910390f35b61008e60025481565b60405190815260200161007c565b6100af6100aa366004610341565b61018a565b005b6100c46100bf36600461036b565b6101d4565b604051901515815260200161007c565b61008e6100e236600461031f565b60036020526000908152604090205481565b61006f6102f6565b60008054610109906104e6565b80601f0160208091040260200160405190810160405280929190818152602001828054610135906104e6565b80156101825780601f1061015757610100808354040283529160200191610182565b820191906000526020600020905b81548152906001019060200180831161016557829003601f168201915b505050505081565b6001600160a01b038216600090815260036020526040812080548392906101b29084906104b7565b9250508190555080600260008282546101cb91906104b7565b90915550505050565b3360009081526003602052604081205483111561022e5760405162461bcd60e51b8152602060048201526014602482015273496e73756666696369656e742062616c616e636560601b604482015260640160405180910390fd5b60005b600681101561029e5733600090815260036020526040812080548692906102599084906104cf565b90915550506001600160a01b038516600090815260036020526040812080548692906102869084906104b7565b9091555081905061029681610521565b915050610231565b50836001600160a01b0316336001600160a01b03167fcd6e659e4c2e75c3bfe47fecaccf39aeb368116a0ee52afb532e07f6cba6c0d185856040516102e4929190610496565b60405180910390a35060019392505050565b60018054610109906104e6565b80356001600160a01b038116811461031a57600080fd5b919050565b60006020828403121561033157600080fd5b61033a82610303565b9392505050565b6000806040838503121561035457600080fd5b61035d83610303565b946020939093013593505050565b60008060006060848603121561038057600080fd5b61038984610303565b925060208401359150604084013567ffffffffffffffff808211156103ad57600080fd5b818601915086601f8301126103c157600080fd5b8135818111156103d3576103d3610552565b604051601f8201601f19908116603f011681019083821181831017156103fb576103fb610552565b8160405282815289602084870101111561041457600080fd5b8260208601602083013760006020848301015280955050505050509250925092565b6000815180845260005b8181101561045c57602081850181015186830182015201610440565b8181111561046e576000602083870101525b50601f01601f19169290920160200192915050565b60208152600061033a6020830184610436565b8281526040602082015260006104af6040830184610436565b949350505050565b600082198211156104ca576104ca61053c565b500190565b6000828210156104e1576104e161053c565b500390565b600181811c908216806104fa57607f821691505b6020821081141561051b57634e487b7160e01b600052602260045260246000fd5b50919050565b60006000198214156105355761053561053c565b5060010190565b634e487b7160e01b600052601160045260246000fd5b634e487b7160e01b600052604160045260246000fdfea26469706673582212205393945b341e6ee5bb89c350656b19b8439867683ac45c09491327bbfba1d64864736f6c63430008060033","type":"CREATE","value":"0x"}}

```
 The same calls on Goerly
 
 
[debug_traceTranscation_goerly.txt](https://github.com/skalenetwork/skaled/files/13707484/debug_traceTranscation_goerly.txt)

### PR DESCRIPTION
**Pull request by kladkogex**
WHAT WAS FIXED

Trace was not working if a block has multiple transactions (from a single entity or from multiple entities) 

Bug 1751 is a duplicate of 1750. Both are fixed by this commit,.

1) When the block was executed, receipts were not saved.   Execution of a transaction requires receipt of the previous transaction since it includes the totalgas used so far. This is important for block gas limit.

2) Also the case when a transaction does not call a smartcontract (so the trace is empty) was not handled properly.

3) In the process of the fix, I made corresponding exceptions more informative and checked that exceptions are always handled

TESTS

I modified the test tracer.ts. Now the test inserts a transaction before the test transaction, so in the test the block 
includes two transactions.

As before, the test runs as part of github actions.

Bug 1750 is a duplicate of 1751. Both are fixed by this commit,.

### CODE REVIEW

File: libethereum/Client.cpp [hunk 1]
```diff
@@ -1346,43 +1346,49 @@ Transaction Client::createTransactionForCallOrTraceCall( const Address& _from, c
...
+        string key = to_string( _blockNumber ) + traceOptions.toString();
 
-    auto cachedResult = m_blockTraceCache.getIfExists( key );
-    if ( cachedResult.has_value() ) {
-        return std::any_cast< Json::Value >( cachedResult );
-    }
+        auto cachedResult = m_blockTraceCache.getIfExists( key );
+        if ( cachedResult.has_value() ) {
+            return std::any_cast< Json::Value >( cachedResult );
+            cerr << "match !!!" << endl;
```
Comment (DmytroNazarenko) [inline libethereum/Client.cpp]: Please, remove redundant log

File: libhistoric/AlethStandardTrace.cpp [hunk 2]
```diff
@@ -361,24 +375,29 @@ string AlethStandardTrace::toGethCompatibleCompactHexPrefixed( const u256& _valu
...
-
+    if ( getTopFunctionCall() ) {
+        recordFunctionReturned( statusCode, _er.output, m_totalGasUsed );
+    }
+    // we are done. Set the trace to finalized
+    STATE_CHECK( !m_isFinalized.exchange( true ) )
+    // now print trace
+    printTrace( _er, _statePre, _statePost );
+}
+void eth::AlethStandardTrace::printTrace( ExecutionResult& _er, const HistoricState& _statePre,
```
Comment (DmytroNazarenko) [inline libhistoric/AlethStandardTrace.cpp:404]: Please, add a blank line between methods

File: libhistoric/AlethStandardTrace.cpp [hunk 3]
```diff
@@ -443,17 +464,32 @@ const u256& AlethStandardTrace::getMinerPayment() const {
...
-    this->m_accessedAccounts.insert( m_blockAuthor );
+    m_accessedAccounts.insert( m_blockAuthor );
 }
 bool AlethStandardTrace::isCall() const {
     return m_isCall;
 }
 const u256& AlethStandardTrace::getOriginalFromBalance() const {
     return m_originalFromBalance;
 }
+
```
Comment (DmytroNazarenko) [inline libhistoric/AlethStandardTrace.cpp:494]: Please, use 1 blank line between methods

File: libhistoric/ReplayTracePrinter.cpp [hunk 4]
```diff
@@ -45,7 +45,13 @@ void ReplayTracePrinter::print( Json::Value& _jsonTrace, const ExecutionResult&
...
     Json::Value emptyAddress( Json::arrayValue );
 
-    m_trace.getTopFunctionCall()->printParityFunctionTrace( functionTraceArray, emptyAddress );
+    auto topFunctionCallRecord = m_trace.getTopFunctionCall();
+    if ( !topFunctionCallRecord ) {
+        // no bytecodes were executed, this was purely ETH transfer
+        // print nothing
+    } else {
+        topFunctionCallRecord->printParityFunctionTrace( functionTraceArray, emptyAddress );
+    }
```
Comment (DmytroNazarenko) [inline libhistoric/ReplayTracePrinter.cpp]: Probably it is better to replace
```
if ( !topFunctionCallRecord ) {
} else {
}
```
with
```
if ( topFunctionCallRecord ) {
}
```

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Review by olehnikolaiev — APPROVED** [review]

**Pull request merged**